X chromosome functionality

AprilYUZhang opened the pull request:
This is `tinyhouse` work to fix X-chromosome treatment in AlphaSuite tools https://github.com/AlphaGenes/AlphaPeel/issues/66

gregorgorjanc committed:
Merge pull request #6 from XingerTang/main

Revert the change of writePhase()

gregorgorjanc committed:
Merge pull request #13 from AlphaGenes/devel

Merging recent devel changes into main

AprilYUZhang committed:
Implement new feature

-Update segregation prob and error matrix for x chromsome
-Update log
-Change all to XChrMaleFlag

gregorgorjanc commented on ProbMath.py:
```diff
@@ -34,16 +34,19 @@ def getGenotypeProbabilities_ind(ind, args = None, log = False):
-
     if genotypes is not None:
-        setGenoProbsFromGenotypes(genotypes, errorMat, vals)
+        if useSexChrom:
```
Please change to `useXChrom` so we can later implement also `useYChrom`.

gregorgorjanc commented on ProbMath.py:
```diff
@@ -34,16 +34,19 @@ def getGenotypeProbabilities_ind(ind, args = None, log = False):
 
 
 def getGenotypeProbabilities(nLoci, genotypes, reads, error = 0.01, seqError = 0.001, useSexChrom=False):
```
Please change to `useXChrom` so we can later implement also `useYChrom`.

AprilYUZhang replied:
Change to XChrMaleFlag

gregorgorjanc commented on ProbMath.py:
```diff
@@ -34,16 +34,19 @@ def getGenotypeProbabilities_ind(ind, args = None, log = False):
     if genotypes is not None:
-        setGenoProbsFromGenotypes(genotypes, errorMat, vals)
+        if useSexChrom:
+            errorMat = generateErrorMat_Xchr_male(error)
```
Why do we do it here just for males? We are in the `if genotypes is not None` loop, but genotype could be both from male or female.

AprilYUZhang replied:
in 24 row 
"sexChromFlag = getattr(args, "sexchrom", False) and ind.sex == 0 #This is the sex chromosome and the individual is male. 
return getGenotypeProbabilities(nLoci, ind.genotypes, ind.reads, error, seqError, sexChromFlag)" 
In fact, this sexChromFlag means the use of sex chromosomes, and the individual is male

AprilYUZhang replied:
I suggest to change sexChromFlag to XChrMaleFlag

gregorgorjanc replied:
ok

gregorgorjanc commented on ProbMath.py:
```diff
@@ -59,11 +62,11 @@ def getGenotypeProbabilities(nLoci, genotypes, reads, error = 0.01, seqError = 0
-        vals[1,:] = vals[3,:]
-        vals[2,:] = 0
-        vals[3,:] = 0
+    # if useSexChrom:
```
I suggest you now remove these lines

gregorgorjanc commented on ProbMath.py:
```diff
@@ -75,11 +78,14 @@ def getGenotypeProbabilities_log(nLoci, genotypes, reads, error = 0.01, seqError
-        setGenoProbsFromGenotypes(genotypes, errorMat, vals)
-    
+        if useSexChrom:
+            errorMat = generateErrorMat_Xchr_male(error)
```
As above

gregorgorjanc commented on ProbMath.py:
```diff
@@ -213,17 +219,31 @@ def setGenoProbsFromGenotypes(genotypes, errorMat, vals):
     for i in range(nLoci) :
         if genotypes[i] != 9:
             vals[:, i] = errorMat[genotypes[i], :, i]
+def setGenoProbsFromGenotypes_Xchr_male(genotypes, errorMat, vals):
```
Add empty line above. Use black

gregorgorjanc commented on ProbMath.py:
```diff
@@ -213,17 +219,31 @@ def setGenoProbsFromGenotypes(genotypes, errorMat, vals):
+            if genotypes[i] == 2:
+                vals[:, i] = np.zeros(4)
+            else:
+                vals[:, i] = errorMat[genotypes[i], :, i]
```
Add empty line below. Use black

gregorgorjanc commented on ProbMath.py:
```diff
@@ -250,48 +263,42 @@ def generateSegregationXXChrom(partial=False, e= 1e-06) :
-def generateSegregation(partial=False, e= 1e-06) :
-    paternalTransmission = np.array([ [1, 1, 0, 0],[0, 0, 1, 1]])
-    maternalTransmission = np.array([ [1, 0, 1, 0],[0, 1, 0, 1]])
+def generateSegregation_mu(partial=False, mu= 1e-08) :
```
Keep the old name

gregorgorjanc commented on ProbMath.py:
```diff
@@ -250,48 +263,42 @@ def generateSegregationXXChrom(partial=False, e= 1e-06) :
         for allele in range(4) :
             segregationTensor[:, :, allele, segregation] = np.outer(father[fatherAlleleCoding[allele]], mother[motherAlleleCoding[allele]])
-
-    segregationTensor = segregationTensor*(1-e) + e/4 #trace has 4 times as many elements as it should since it has 4 internal reps.
```
Can you leave this line in but comment it out, just in case we might need it later for numerical reasons, but I see that using `mu` will now solve this problem.

gregorgorjanc commented on ProbMath.py:
```diff
@@ -320,7 +327,6 @@ def generateSegregation(partial=False, e= 1e-06) :
 
     if partial : segregationTensor = np.mean(segregationTensor, 3)
 
-    segregationTensor = segregationTensor*(1-e) + e/4 #trace has 4 times as many elements as it should since it has 4 internal reps.
```
As above - comment out

gregorgorjanc reviewed (commented):
See comments - only minor bits and we merge this in.

Also, provide a CHANGELOG line;)

gregorgorjanc commented on CHANGELOG.md:
```diff
@@ -6,6 +6,8 @@
 
 - Added metafounders to pedigree ({pr}`23`, {user}`RosCraddock`, {user}`XingerTang`)
 
+- fix X chromosome's generateSegregation and errorMatrix ({pr}`24`, {user}`AprilYUZhang`)
```
fix --> Fix (so we start each sentence with capital letter)

gregorgorjanc reviewed (commented):
Nearly there @AprilYUZhang ;) One more fix and we are ready to merge. Please also squash commits.

AprilYUZhang commented:
@gregorgorjanc Squashed!

AprilYUZhang committed:
Fix typo in CHANGELOG for X chromosome entry

Corrected formatting for X chromosome's generateSegregation and errorMatrix entry.

XingerTang commented on ProbMath.py:
```diff
@@ -166,179 +190,237 @@ def set_from_genotype_probs(ind, geno_probs = None, calling_threshold = 0.1, set
-        for allele in range(4) :
-            segregationTensor[:, :, allele, segregation] = np.outer(father[fatherAlleleCoding[allele]], mother[motherAlleleCoding[allele]])
-
-    segregationTensor = segregationTensor*(1-e) + e/4 #trace has 4 times as many elements as it should since it has 4 internal reps.
```
If I didn't get it wrong, is that the ``generateSegregationXXChrom`` exactly the same as the regular ``generateSegregation``? If that is the case, maybe we can only keep one.

gregorgorjanc replied:
Hmm, I am blanking out just now. @AprilYUZhang please meet with @XingerTang and discuss the differences and let me know if you guys need my input. @XingerTang we can also look at this tomorrow in our meeting.

AprilYUZhang replied:
Because we know this X and X, so we can make sure segregation 0 and 1 (father's father mustn't contribute to the offspring) has no possible occurrence, this is the biggest difference.

XingerTang replied:
Thank you for the clarification. I didn't notice that part. 

To make this more efficient, could you edit lines 316 to 322 as:
```
father = maternalTransmission
for segregation in range(2, 4):
  if segregation == 2:
    mother = paternalTransmission
  elif segregation == 3:
    mother = maternalTransmission
```

AprilYUZhang replied:
here it is [cc806f7](https://github.com/AlphaGenes/tinyhouse/pull/24/commits/cc806f70a46bd070c0ab9c84f497147887c6c0c2)

XingerTang reviewed (changes requested):
Looks all good to me except the comment.

AprilYUZhang committed:
Merge branch 'AlphaGenes:main' into sex_chr

AprilYUZhang committed:
use warning when  male genotype coded as '2'

AprilYUZhang committed:
add warning message

AprilYUZhang committed:
Efficient edit

XingerTang commented on ProbMath.py:
```diff
@@ -314,11 +314,10 @@ def generateSegregationXXChrom(partial=False, mu=1e-08):
 
     segregationTensor = np.zeros((4, 4, 4, 4))
     for segregation in range(2, 4):
+        father = maternalTransmission
```
It can be put before the loop, so there are no multiple assignments.

XingerTang replied:
@AprilYUZhang You might have missed this.

AprilYUZhang replied:
[fix](https://github.com/AlphaGenes/tinyhouse/pull/24/commits/3e1b8a5f241b2de8d90cbff667000cfd2432fbeb)
Here it is.

AprilYUZhang committed:
fix
